src/mgr/PyModuleRegistry.h

```diff
@@ -81,26 +81,42 @@ class PyModuleRegistry
     : clog(clog_)
   {}
 
+  /**
+   * @return true if the mgrmap has changed such that the service needs restart
+   */
   bool handle_mgr_map(const MgrMap &mgr_map_)
   {
     Mutex::Locker l(lock);
 
-    bool modules_changed = mgr_map_.modules != mgr_map.modules;
-    mgr_map = mgr_map_;
-
-    if (standby_modules != nullptr) {
-      standby_modules->handle_mgr_map(mgr_map_);
+    if (mgr_map.epoch == 0) {
+      // First time we see MgrMap, set the enabled flags on modules
+      // This should always happen before someone calls standby_start
+      // or active_start
+      for (const auto &i : modules) {
+        const auto &module = i.second;
+        const bool enabled = (mgr_map.modules.count(module_name) > 0);
```

Comment:
```
/home/jenkins-build/build/workspace/ceph-dev-new-build/ARCH/x86_64/AVAILABLE_ARCH/x86_64/AVAILABLE_DIST/centos7/DIST/centos7/MACHINE_SIZE/huge/release/13.0.1-2432-g7c5dfdf/rpm/el7/BUILD/ceph-13.0.1-2432-g7c5dfdf/src/mgr/PyModuleRegistry.h:97:53: error: 'module_name' was not declared in this scope
         const bool enabled = (mgr_map.modules.count(module_name) > 0);
                                                     ^~~~~~~~~~~
```

Reply:
could also put following, as we have structured binding in C++17 now:
```c++
for (auto& [module_name, module] : modules) {
  const bool enabled = (mgr_map.modules.count(module_name) > 0);
  module.set_enabled(enabled);
}
```

please note `PyModule::set_enabled()` is not a const method. so we cannot use `const auto& [module_name, module]` here.

Reply:
Thanks for suggestions, this is my first time using the new structured binding bits and I like it :-)

The `const auto` is permitted (if a bit misleading) because the `module` part is a shared_ptr rather than the actual PyModule

I've pushed an update.